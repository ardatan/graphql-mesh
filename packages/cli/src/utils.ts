export const stripJSONComments = (data: string) => {
  const re = /\/\/(.*)/g;
  const re2 = /\/\*(.*)\*\//g;
  return data.replace(re, '').replace(re2, '');
};
