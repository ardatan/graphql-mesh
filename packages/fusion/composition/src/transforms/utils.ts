export class TransformValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'TransformValidationError';
  }
}
