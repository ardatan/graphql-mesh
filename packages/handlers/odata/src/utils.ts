export function getUrlString(url: URL) {
  return decodeURIComponent(url.toString()).split('+').join(' ');
}

export function addIdentifierToUrl(url: URL, identifierFieldName: string, identifierFieldTypeRef: string, args: any) {
  url.href += `/${args[identifierFieldName]}/`;
}
