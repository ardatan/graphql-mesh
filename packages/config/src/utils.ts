import { parse } from 'graphql';
import { MeshHandlerLibrary, KeyValueCache, YamlConfig, MergerFn, ImportFn, MeshPubSub } from '@graphql-mesh/types';
import { resolve } from 'path';
import { IResolvers, printSchemaWithDirectives } from '@graphql-tools/utils';
import { paramCase } from 'param-case';
import { loadTypedefs } from '@graphql-tools/load';
import { GraphQLFileLoader } from '@graphql-tools/graphql-file-loader';
import { get, set, kebabCase } from 'lodash';
import { stringInterpolator } from '@graphql-mesh/utils';
import { mergeResolvers } from '@graphql-tools/merge';
import { PubSub, withFilter } from 'graphql-subscriptions';
import { EventEmitter } from 'events';
import { CodeFileLoader } from '@graphql-tools/code-file-loader';
import StitchingMerger from '@graphql-mesh/merger-stitching';

export async function getPackage<T>(name: string, type: string, importFn: ImportFn): Promise<T> {
  const casedName = paramCase(name);
  const casedType = paramCase(type);
  const possibleNames = [
    `@graphql-mesh/${casedName}`,
    `@graphql-mesh/${casedName}-${casedType}`,
    `@graphql-mesh/${casedType}-${casedName}`,
    casedName,
    `${casedName}-${casedType}`,
    `${casedType}-${casedName}`,
    casedType,
  ];
  if (name.includes('-')) {
    possibleNames.push(name);
  }
  const possibleModules = possibleNames.concat(resolve(process.cwd(), name));

  for (const moduleName of possibleModules) {
    try {
      const exported = await importFn(moduleName);

      return (exported.default || exported.parser || exported) as T;
    } catch (err) {
      if (
        !err.message.includes(`Cannot find module '${moduleName}'`) &&
        !err.message.includes(`Could not locate module`)
      ) {
        throw new Error(`Unable to load ${type} matching ${name}: ${err.message}`);
      }
    }
  }

  throw new Error(`Unable to find ${type} matching ${name}`);
}

export async function getHandler(name: keyof YamlConfig.Handler, importFn: ImportFn): Promise<MeshHandlerLibrary> {
  const handlerFn = await getPackage<MeshHandlerLibrary>(name.toString(), 'handler', importFn);

  return handlerFn;
}

export async function resolveAdditionalTypeDefs(baseDir: string, additionalTypeDefs: string) {
  if (additionalTypeDefs) {
    const sources = await loadTypedefs(additionalTypeDefs, {
      cwd: baseDir,
      loaders: [new CodeFileLoader(), new GraphQLFileLoader()],
    });
    return sources.map(source => source.document || parse(source.rawSDL || printSchemaWithDirectives(source.schema)));
  }
  return undefined;
}

export async function resolveAdditionalResolvers(
  baseDir: string,
  additionalResolvers: (
    | string
    | YamlConfig.AdditionalStitchingResolverObject
    | YamlConfig.AdditionalSubscriptionObject
  )[],
  importFn: ImportFn,
  pubsub: MeshPubSub
): Promise<IResolvers> {
  const loadedResolvers = await Promise.all(
    (additionalResolvers || []).map(async additionalResolver => {
      if (typeof additionalResolver === 'string') {
        const filePath = additionalResolver;

        const exported = await importFn(resolve(baseDir, filePath));
        let resolvers = null;

        if (exported.default) {
          if (exported.default.resolvers) {
            resolvers = exported.default.resolvers;
          } else if (typeof exported.default === 'object') {
            resolvers = exported.default;
          }
        } else if (exported.resolvers) {
          resolvers = exported.resolvers;
        }

        if (!resolvers) {
          console.warn(`Unable to load resolvers from file: ${filePath}`);

          return {};
        }

        return resolvers;
      } else {
        if ('pubsubTopic' in additionalResolver) {
          return {
            [additionalResolver.type]: {
              [additionalResolver.field]: {
                subscribe: withFilter(
                  (root, args, context, info) => {
                    const resolverData = { root, args, context, info };
                    const topic = stringInterpolator.parse(additionalResolver.pubsubTopic, resolverData);
                    return pubsub.asyncIterator(topic);
                  },
                  (root, args, context, info) => {
                    return additionalResolver.filterBy ? eval(additionalResolver.filterBy) : true;
                  }
                ),
                resolve: (payload: any) => {
                  if (additionalResolver.returnData) {
                    return get(payload, additionalResolver.returnData);
                  }
                  return payload;
                },
              },
            },
          };
        } else {
          return {
            [additionalResolver.type]: {
              [additionalResolver.field]: {
                selectionSet: additionalResolver.requiredSelectionSet,
                resolve: async (root: any, args: any, context: any, info: any) => {
                  const resolverData = { root, args, context, info };
                  const methodArgs: any = {};
                  for (const argPath in additionalResolver.args) {
                    set(methodArgs, argPath, stringInterpolator.parse(additionalResolver.args[argPath], resolverData));
                  }
                  const result = await context[additionalResolver.targetSource].api[additionalResolver.targetMethod](
                    methodArgs,
                    {
                      selectedFields: additionalResolver.resultSelectedFields,
                      selectionSet: additionalResolver.resultSelectionSet,
                      depth: additionalResolver.resultDepth,
                    }
                  );
                  return additionalResolver.returnData ? get(result, additionalResolver.returnData) : result;
                },
              },
            },
          };
        }
      }
    })
  );

  return mergeResolvers(loadedResolvers);
}

export async function resolveCache(
  cacheConfig: YamlConfig.Config['cache'],
  importFn: ImportFn
): Promise<KeyValueCache | undefined> {
  if (cacheConfig) {
    const cacheName = Object.keys(cacheConfig)[0];
    const config = cacheConfig[cacheName];

    const moduleName = kebabCase(cacheName.toString());
    const pkg = await getPackage<any>(moduleName, 'cache', importFn);
    const Cache = pkg.default || pkg;

    return new Cache(config);
  }
  const InMemoryLRUCache = await import('@graphql-mesh/cache-inmemory-lru').then(m => m.default);
  const cache = new InMemoryLRUCache();
  return cache;
}

export async function resolvePubSub(
  pubsubYamlConfig: YamlConfig.Config['pubsub'],
  importFn: ImportFn
): Promise<MeshPubSub> {
  if (pubsubYamlConfig) {
    let pubsubName: string;
    let pubsubConfig: any;
    if (typeof pubsubYamlConfig === 'string') {
      pubsubName = pubsubYamlConfig;
    } else {
      pubsubName = pubsubYamlConfig.name;
      pubsubConfig = pubsubYamlConfig.config;
    }

    const moduleName = kebabCase(pubsubName.toString());
    const pkg = await getPackage<any>(moduleName, 'pubsub', importFn);
    const PubSub = pkg.default || pkg;

    return new PubSub(pubsubConfig);
  } else {
    const eventEmitter = new EventEmitter({ captureRejections: true });
    eventEmitter.setMaxListeners(Infinity);
    const pubsub = new PubSub({ eventEmitter }) as MeshPubSub;

    return pubsub;
  }
}

export async function resolveMerger(mergerConfig: YamlConfig.Config['merger'], importFn: ImportFn): Promise<MergerFn> {
  if (mergerConfig) {
    const pkg = await getPackage<any>(mergerConfig, 'merger', importFn);
    return pkg.default || pkg;
  }
  return StitchingMerger;
}
