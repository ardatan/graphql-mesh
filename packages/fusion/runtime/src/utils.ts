import { print, type DocumentNode, type ExecutionResult, type GraphQLSchema } from 'graphql';
import type { TransportOptions } from '@graphql-mesh/serve-runtime';
import type {
  Transport,
  TransportContext,
  TransportEntry,
  TransportGetSubgraphExecutor,
  TransportGetSubgraphExecutorOptions,
} from '@graphql-mesh/transport-common';
import type { Logger } from '@graphql-mesh/types';
import { iterateAsync, mapMaybePromise } from '@graphql-mesh/utils';
import {
  isAsyncIterable,
  isDocumentNode,
  mapAsyncIterator,
  printSchemaWithDirectives,
  type ExecutionRequest,
  type Executor,
  type Maybe,
  type MaybePromise,
} from '@graphql-tools/utils';

export type { TransportEntry, TransportGetSubgraphExecutor, TransportGetSubgraphExecutorOptions };

export type Transports =
  | {
      [Kind in string]: MaybePromise<
        | Transport<Kind> // named export
        | {
            default: Transport<Kind>; // property on default export
          }
      >;
    }
  | (<Kind extends string>(
      kind: Kind,
    ) => MaybePromise<
      | Transport<Kind> // named export
      | {
          default: Transport<Kind>; // property on default export
        }
    >);

export type TransportExecutorFactoryGetter = (
  kind: string,
) => MaybePromise<TransportGetSubgraphExecutor>;

function createTransportExecutorFactoryGetter(
  transports: Transports,
  transportOptions: TransportOptions,
  logger: Logger,
): TransportExecutorFactoryGetter {
  return async function getTransport(kind) {
    let transportGetSubgraphExecutor: TransportGetSubgraphExecutor;
    let transport =
      typeof transports === 'function' ? await transports(kind) : await transports?.[kind];
    if (transport) {
      transportGetSubgraphExecutor =
        'default' in transport
          ? transport.default.getSubgraphExecutor
          : transport.getSubgraphExecutor;
    } else {
      const childLogger = logger.child?.(kind);
      try {
        transport = await import(`@graphql-mesh/transport-${kind}`);
      } catch (err) {
        childLogger.error(err);
        throw new Error(
          `No transport found for ${kind}. Please make sure you have installed @graphql-mesh/transport-${kind} or defined the transport config in "mesh.config.ts"`,
        );
      }
      transportGetSubgraphExecutor =
        'default' in transport
          ? transport.default.getSubgraphExecutor
          : transport.getSubgraphExecutor;
      if (!transportGetSubgraphExecutor) {
        throw new Error(
          `@graphql-mesh/transport-${kind} module does not export "getSubgraphExecutor"`,
        );
      }
      if (typeof transportGetSubgraphExecutor !== 'function') {
        throw new Error(
          `@graphql-mesh/transport-${kind} module's export "getSubgraphExecutor" is not a function`,
        );
      }
    }
    return function getSubgraphExecutor(opts) {
      return transportGetSubgraphExecutor({
        ...opts,
        transportEntry: {
          ...opts.transportEntry,
          options: {
            ...opts.transportEntry.options,
            ...transportOptions?.['*']?.[kind],
            ...transportOptions?.[opts.transportEntry.subgraph]?.[kind],
          },
        },
      });
    };
  };
}

function getTransportExecutor(
  transportExecutorFactoryGetter: TransportExecutorFactoryGetter,
  transportContext: TransportGetSubgraphExecutorOptions,
  transportExecutorStack: AsyncDisposableStack,
): MaybePromise<Executor> {
  const kind = transportContext.transportEntry?.kind || '';
  const subgraphName = transportContext.subgraphName || '';
  transportContext.logger?.info(`Loading transport ${kind} for subgraph ${subgraphName}`);
  return mapMaybePromise(transportExecutorFactoryGetter(kind), executor =>
    mapMaybePromise(executor(transportContext), executor => {
      if (isDisposable(executor)) {
        transportExecutorStack.use(executor);
      }
      return executor;
    }),
  );
}

/**
 * This function creates a executor factory that uses the transport packages,
 * and wraps them with the hooks
 */
export function getOnSubgraphExecute({
  onSubgraphExecuteHooks,
  transportContext,
  transportEntryMap,
  getSubgraphSchema,
  transportExecutorStack,
  transports,
  transportOptions,
}: {
  onSubgraphExecuteHooks: OnSubgraphExecuteHook[];
  transports?: Transports;
  transportOptions?: TransportOptions;
  transportContext?: TransportContext;
  transportEntryMap?: Record<string, TransportEntry>;
  getSubgraphSchema(subgraphName: string): GraphQLSchema;
  transportExecutorStack: AsyncDisposableStack;
}) {
  const subgraphExecutorMap = new Map<string, Executor>();
  const transportExecutorFactoryGetter = createTransportExecutorFactoryGetter(
    transports,
    transportOptions,
    transportContext.logger,
  );
  return function onSubgraphExecute(subgraphName: string, executionRequest: ExecutionRequest) {
    let executor: Executor = subgraphExecutorMap.get(subgraphName);
    // If the executor is not initialized yet, initialize it
    if (executor == null) {
      transportContext?.logger?.info(`Initializing executor for subgraph ${subgraphName}`);
      // Lazy executor that loads transport executor on demand
      executor = function lazyExecutor(subgraphExecReq: ExecutionRequest) {
        return mapMaybePromise(
          // Gets the transport executor for the given subgraph
          getTransportExecutor(
            transportExecutorFactoryGetter,
            transportContext
              ? {
                  ...transportContext,
                  subgraphName,
                  get subgraph() {
                    return getSubgraphSchema(subgraphName);
                  },
                  get transportEntry() {
                    return transportEntryMap?.[subgraphName];
                  },
                }
              : {
                  get subgraph() {
                    return getSubgraphSchema(subgraphName);
                  },
                  get transportEntry() {
                    return transportEntryMap?.[subgraphName];
                  },
                  subgraphName,
                },
            transportExecutorStack,
          ),
          executor_ => {
            // Wraps the transport executor with hooks
            executor = wrapExecutorWithHooks({
              executor: executor_,
              onSubgraphExecuteHooks,
              subgraphName,
              transportEntryMap,
              getSubgraphSchema,
            });
            // Caches the executor for future use
            subgraphExecutorMap.set(subgraphName, executor);
            return executor(subgraphExecReq);
          },
        );
      };
      // Caches the lazy executor to prevent race conditions
      subgraphExecutorMap.set(subgraphName, executor);
    }
    return executor(executionRequest);
  };
}

export interface WrapExecuteWithHooksOptions {
  executor: Executor;
  onSubgraphExecuteHooks: OnSubgraphExecuteHook[];
  subgraphName: string;
  transportEntryMap?: Record<string, TransportEntry>;
  getSubgraphSchema: (subgraphName: string) => GraphQLSchema;
}

/**
 * This function wraps the executor created by the transport package
 * with `onSubgraphExecuteHooks` to hook into the execution phase of subgraphs
 */
export function wrapExecutorWithHooks({
  executor,
  onSubgraphExecuteHooks,
  subgraphName,
  transportEntryMap,
  getSubgraphSchema,
}: WrapExecuteWithHooksOptions): Executor {
  if (onSubgraphExecuteHooks.length === 0) {
    return executor;
  }
  return function executorWithHooks(executionRequest: ExecutionRequest) {
    const onSubgraphExecuteDoneHooks: OnSubgraphExecuteDoneHook[] = [];
    return mapMaybePromise(
      iterateAsync(
        onSubgraphExecuteHooks,
        onSubgraphExecuteHook =>
          onSubgraphExecuteHook({
            get subgraph() {
              return getSubgraphSchema(subgraphName);
            },
            subgraphName,
            get transportEntry() {
              return transportEntryMap?.[subgraphName];
            },
            executionRequest,
            setExecutionRequest(newExecutionRequest) {
              executionRequest = newExecutionRequest;
            },
            executor,
            setExecutor(newExecutor) {
              executor = newExecutor;
            },
          }),
        onSubgraphExecuteDoneHooks,
      ),
      () => {
        if (onSubgraphExecuteDoneHooks.length === 0) {
          return executor(executionRequest);
        }
        return mapMaybePromise(executor(executionRequest), currentResult => {
          const executeDoneResults: OnSubgraphExecuteDoneResult[] = [];
          return mapMaybePromise(
            iterateAsync(
              onSubgraphExecuteDoneHooks,
              onSubgraphExecuteDoneHook =>
                onSubgraphExecuteDoneHook({
                  result: currentResult,
                  setResult(newResult: ExecutionResult) {
                    currentResult = newResult;
                  },
                }),
              executeDoneResults,
            ),
            () => {
              if (!isAsyncIterable(currentResult)) {
                return currentResult;
              }

              if (executeDoneResults.length === 0) {
                return currentResult;
              }

              const onNextHooks: OnSubgraphExecuteDoneResultOnNext[] = [];
              const onEndHooks: OnSubgraphExecuteDoneResultOnEnd[] = [];

              for (const executeDoneResult of executeDoneResults) {
                if (executeDoneResult.onNext) {
                  onNextHooks.push(executeDoneResult.onNext);
                }
                if (executeDoneResult.onEnd) {
                  onEndHooks.push(executeDoneResult.onEnd);
                }
              }

              if (onNextHooks.length === 0 && onEndHooks.length === 0) {
                return currentResult;
              }

              const asyncIterator = currentResult[Symbol.asyncIterator]();
              return mapAsyncIterator(
                asyncIterator,
                currentResult =>
                  mapMaybePromise(
                    iterateAsync(onNextHooks, onNext =>
                      onNext({
                        result: currentResult,
                        setResult: res => {
                          currentResult = res;
                        },
                      }),
                    ),
                    () => currentResult,
                  ),
                undefined,
                () =>
                  onEndHooks.length === 0 ? undefined : iterateAsync(onEndHooks, onEnd => onEnd()),
              );
            },
          );
        });
      },
    );
  };
}

export interface UnifiedGraphPlugin {
  onSubgraphExecute?: OnSubgraphExecuteHook;
}

export type OnSubgraphExecuteHook = (
  payload: OnSubgraphExecutePayload,
) => Promise<Maybe<OnSubgraphExecuteDoneHook | void>> | Maybe<OnSubgraphExecuteDoneHook | void>;

export interface OnSubgraphExecutePayload {
  subgraph: GraphQLSchema;
  subgraphName: string;
  transportEntry?: TransportEntry;
  executionRequest: ExecutionRequest;
  setExecutionRequest(executionRequest: ExecutionRequest): void;
  executor: Executor;
  setExecutor(executor: Executor): void;
}

export interface OnSubgraphExecuteDonePayload {
  result: AsyncIterable<ExecutionResult> | ExecutionResult;
  setResult(result: AsyncIterable<ExecutionResult> | ExecutionResult): void;
}

export type OnSubgraphExecuteDoneHook = (
  payload: OnSubgraphExecuteDonePayload,
) => MaybePromise<Maybe<OnSubgraphExecuteDoneResult | void>>;

export type OnSubgraphExecuteDoneResultOnNext = (
  payload: OnSubgraphExecuteDoneOnNextPayload,
) => MaybePromise<void>;

export interface OnSubgraphExecuteDoneOnNextPayload {
  result: ExecutionResult;
  setResult(result: ExecutionResult): void;
}

export type OnSubgraphExecuteDoneResultOnEnd = () => MaybePromise<void>;

export type OnSubgraphExecuteDoneResult = {
  onNext?: OnSubgraphExecuteDoneResultOnNext;
  onEnd?: OnSubgraphExecuteDoneResultOnEnd;
};

export function compareSchemas(
  a: DocumentNode | string | GraphQLSchema,
  b: DocumentNode | string | GraphQLSchema,
) {
  let aStr: string;
  if (typeof a === 'string') {
    aStr = a;
  } else if (isDocumentNode(a)) {
    aStr = print(a);
  } else {
    aStr = printSchemaWithDirectives(a);
  }
  let bStr: string;
  if (typeof b === 'string') {
    bStr = b;
  } else if (isDocumentNode(b)) {
    bStr = print(b);
  } else {
    bStr = printSchemaWithDirectives(b);
  }
  return aStr === bStr;
}

export function isDisposable(obj: any): obj is Disposable | AsyncDisposable {
  return obj?.[Symbol.dispose] != null || obj?.[Symbol.asyncDispose] != null;
}
