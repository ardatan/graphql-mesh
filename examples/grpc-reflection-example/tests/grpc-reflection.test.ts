import { join } from 'path';
import { readFile } from 'fs-extra';
import { findAndParseConfig } from '@graphql-mesh/cli';
import { getMesh, MeshInstance } from '@graphql-mesh/runtime';
import { printSchemaWithDirectives } from '@graphql-tools/utils';
import { Server } from '@grpc/grpc-js';
import startGrpcServer from '../start-server';

describe('gRPC Reflection Example', () => {
  let mesh: MeshInstance;
  let grpc: Server;
  beforeAll(async () => {
    const config = await findAndParseConfig({
      dir: join(__dirname, '..'),
    });
    grpc = await startGrpcServer();
    mesh = await getMesh(config);
  });
  it('should generate correct schema', async () => {
    expect(printSchemaWithDirectives(mesh.schema)).toMatchSnapshot('grpc-schema');
  });
  it('should get greeting correctly', async () => {
    const GreetingQuery = await readFile(
      join(__dirname, '../example-queries/GetGreeting.query.graphql'),
      'utf8',
    );
    const result = await mesh.execute(GreetingQuery);
    expect(result).toMatchSnapshot('greeting-result');
  });
  afterAll(() => {
    mesh?.destroy();
    return new Promise<void>((resolve, reject) => {
      grpc?.tryShutdown(err => {
        if (err) {
          reject(err);
        } else {
          setTimeout(resolve, 1000);
        }
      });
    });
  });
});
