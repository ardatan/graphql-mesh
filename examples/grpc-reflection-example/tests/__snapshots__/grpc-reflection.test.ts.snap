// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gRPC Reflection Example should generate correct schema: grpc-schema 1`] = `
"schema @transport(subgraph: "grpc", kind: "grpc", location: "localhost:50052", options: {requestTimeout: 200000}) {
  query: Query
}

directive @grpcMethod(subgraph: String, rootJsonName: String, objPath: String, methodName: String, responseStream: Boolean) on FIELD_DEFINITION

directive @grpcConnectivityState(subgraph: String, rootJsonName: String, objPath: String) on FIELD_DEFINITION

directive @grpcRootJson(subgraph: String, name: String, rootJson: ObjMap, loadOptions: ObjMap) repeatable on OBJECT

"""
Directs the executor to stream plural fields when the \`if\` argument is true or undefined.
"""
directive @stream(
  """Stream when true or undefined."""
  if: Boolean! = true
  """Unique name"""
  label: String
  """Number of items to return immediately"""
  initialCount: Int = 0
) on FIELD

type Query @grpcRootJson(subgraph: "grpc", name: "Root0", rootJson: "{\\"nested\\":{\\"helloworld\\":{\\"nested\\":{\\"GreetingRequest\\":{\\"fields\\":{\\"name\\":{\\"type\\":\\"string\\",\\"id\\":1}},\\"comment\\":null},\\"GreetingResult\\":{\\"fields\\":{\\"greeting\\":{\\"type\\":\\"string\\",\\"id\\":1}},\\"comment\\":null},\\"GreetingService\\":{\\"methods\\":{\\"GetGreeting\\":{\\"requestType\\":\\".helloworld.GreetingRequest\\",\\"responseType\\":\\".helloworld.GreetingResult\\"}},\\"comment\\":null}}}}}") @grpcRootJson(subgraph: "grpc", name: "Root1", rootJson: "{\\"nested\\":{\\"artnet\\":{\\"nested\\":{\\"coredata\\":{\\"nested\\":{\\"playground\\":{\\"nested\\":{\\"EchoInputRequest\\":{\\"fields\\":{\\"price\\":{\\"type\\":\\"PlayGroundPriceFilter\\",\\"id\\":3}},\\"comment\\":null},\\"EchoUserInfoRequest\\":{\\"fields\\":{},\\"comment\\":null},\\"EchoUserInfoResponse\\":{\\"fields\\":{\\"userInfo\\":{\\"type\\":\\"string\\",\\"id\\":1}},\\"comment\\":null},\\"PlayGroundPriceFilter\\":{\\"fields\\":{\\"range\\":{\\"type\\":\\"PlayGroundPriceRange\\",\\"id\\":1}},\\"comment\\":null},\\"PlayGroundPriceRange\\":{\\"fields\\":{\\"valueDouble\\":{\\"type\\":\\"google.protobuf.DoubleValue\\",\\"id\\":1},\\"valueFloat\\":{\\"type\\":\\"google.protobuf.FloatValue\\",\\"id\\":2},\\"valueLong\\":{\\"type\\":\\"google.protobuf.Int64Value\\",\\"id\\":3},\\"valueInt\\":{\\"type\\":\\"google.protobuf.Int64Value\\",\\"id\\":4},\\"valueLongNotnull\\":{\\"type\\":\\"int64\\",\\"id\\":5}},\\"comment\\":null},\\"SnailMailRequest\\":{\\"fields\\":{\\"allowCancellation\\":{\\"type\\":\\"bool\\",\\"id\\":1},\\"fakeDataLength\\":{\\"type\\":\\"int32\\",\\"id\\":2},\\"waitSeconds\\":{\\"type\\":\\"int32\\",\\"id\\":3}},\\"comment\\":null},\\"SnailMailResponse\\":{\\"fields\\":{\\"fakeData\\":{\\"type\\":\\"string\\",\\"id\\":1}},\\"comment\\":null},\\"Playground\\":{\\"methods\\":{\\"GetInput\\":{\\"requestType\\":\\".artnet.coredata.playground.EchoInputRequest\\",\\"responseType\\":\\".artnet.coredata.playground.EchoInputRequest\\"},\\"GetSnailMail\\":{\\"requestType\\":\\".artnet.coredata.playground.SnailMailRequest\\",\\"responseType\\":\\".artnet.coredata.playground.SnailMailResponse\\"},\\"GetUserInfo\\":{\\"requestType\\":\\".artnet.coredata.playground.EchoUserInfoRequest\\",\\"responseType\\":\\".artnet.coredata.playground.EchoUserInfoResponse\\"}},\\"comment\\":null}}}}}}},\\"google\\":{\\"nested\\":{\\"protobuf\\":{\\"nested\\":{\\"DoubleValue\\":{\\"fields\\":{\\"value\\":{\\"type\\":\\"double\\",\\"id\\":1}},\\"comment\\":null},\\"FloatValue\\":{\\"fields\\":{\\"value\\":{\\"type\\":\\"float\\",\\"id\\":1}},\\"comment\\":null},\\"Int64Value\\":{\\"fields\\":{\\"value\\":{\\"type\\":\\"int64\\",\\"id\\":1}},\\"comment\\":null},\\"UInt64Value\\":{\\"fields\\":{\\"value\\":{\\"type\\":\\"uint64\\",\\"id\\":1}},\\"comment\\":null},\\"Int32Value\\":{\\"fields\\":{\\"value\\":{\\"type\\":\\"int32\\",\\"id\\":1}},\\"comment\\":null},\\"UInt32Value\\":{\\"fields\\":{\\"value\\":{\\"type\\":\\"uint32\\",\\"id\\":1}},\\"comment\\":null},\\"BoolValue\\":{\\"fields\\":{\\"value\\":{\\"type\\":\\"bool\\",\\"id\\":1}},\\"comment\\":null},\\"StringValue\\":{\\"fields\\":{\\"value\\":{\\"type\\":\\"string\\",\\"id\\":1}},\\"comment\\":null},\\"BytesValue\\":{\\"fields\\":{\\"value\\":{\\"type\\":\\"bytes\\",\\"id\\":1}},\\"comment\\":null}}}}}}}") {
  helloworld_GreetingService_GetGreeting(input: helloworld__GreetingRequest_Input): helloworld__GreetingResult @grpcMethod(subgraph: "grpc", rootJsonName: "Root0", objPath: "helloworld.GreetingService", methodName: "GetGreeting", responseStream: false)
  helloworld_GreetingService_connectivityState(tryToConnect: Boolean): ConnectivityState @grpcConnectivityState(subgraph: "grpc", rootJsonName: "Root0", objPath: "helloworld.GreetingService")
  artnet_coredata_playground_Playground_GetInput(input: artnet__coredata__playground__EchoInputRequest_Input): artnet__coredata__playground__EchoInputRequest @grpcMethod(subgraph: "grpc", rootJsonName: "Root1", objPath: "artnet.coredata.playground.Playground", methodName: "GetInput", responseStream: false)
  artnet_coredata_playground_Playground_GetSnailMail(input: artnet__coredata__playground__SnailMailRequest_Input): artnet__coredata__playground__SnailMailResponse @grpcMethod(subgraph: "grpc", rootJsonName: "Root1", objPath: "artnet.coredata.playground.Playground", methodName: "GetSnailMail", responseStream: false)
  artnet_coredata_playground_Playground_GetUserInfo(input: artnet__coredata__playground__EchoUserInfoRequest_Input): artnet__coredata__playground__EchoUserInfoResponse @grpcMethod(subgraph: "grpc", rootJsonName: "Root1", objPath: "artnet.coredata.playground.Playground", methodName: "GetUserInfo", responseStream: false)
  artnet_coredata_playground_Playground_connectivityState(tryToConnect: Boolean): ConnectivityState @grpcConnectivityState(subgraph: "grpc", rootJsonName: "Root1", objPath: "artnet.coredata.playground.Playground")
}

type helloworld__GreetingResult {
  greeting: String
}

input helloworld__GreetingRequest_Input {
  name: String
}

enum ConnectivityState {
  IDLE
  CONNECTING
  READY
  TRANSIENT_FAILURE
  SHUTDOWN
}

type artnet__coredata__playground__EchoInputRequest {
  price: artnet__coredata__playground__PlayGroundPriceFilter
}

type artnet__coredata__playground__PlayGroundPriceFilter {
  range: artnet__coredata__playground__PlayGroundPriceRange
}

type artnet__coredata__playground__PlayGroundPriceRange {
  valueDouble: google__protobuf__DoubleValue
  valueFloat: google__protobuf__FloatValue
  valueLong: google__protobuf__Int64Value
  valueInt: google__protobuf__Int64Value
  valueLongNotnull: BigInt
}

type google__protobuf__DoubleValue {
  value: Float
}

type google__protobuf__FloatValue {
  value: Float
}

type google__protobuf__Int64Value {
  value: BigInt
}

"""
The \`BigInt\` scalar type represents non-fractional signed whole numeric values.
"""
scalar BigInt

input artnet__coredata__playground__EchoInputRequest_Input {
  price: artnet__coredata__playground__PlayGroundPriceFilter_Input
}

input artnet__coredata__playground__PlayGroundPriceFilter_Input {
  range: artnet__coredata__playground__PlayGroundPriceRange_Input
}

input artnet__coredata__playground__PlayGroundPriceRange_Input {
  valueDouble: google__protobuf__DoubleValue_Input
  valueFloat: google__protobuf__FloatValue_Input
  valueLong: google__protobuf__Int64Value_Input
  valueInt: google__protobuf__Int64Value_Input
  valueLongNotnull: BigInt
}

input google__protobuf__DoubleValue_Input {
  value: Float
}

input google__protobuf__FloatValue_Input {
  value: Float
}

input google__protobuf__Int64Value_Input {
  value: BigInt
}

type artnet__coredata__playground__SnailMailResponse {
  fakeData: String
}

input artnet__coredata__playground__SnailMailRequest_Input {
  allowCancellation: Boolean
  fakeDataLength: Int
  waitSeconds: Int
}

type artnet__coredata__playground__EchoUserInfoResponse {
  userInfo: String
}

scalar artnet__coredata__playground__EchoUserInfoRequest_Input @specifiedBy(url: "http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf")

scalar ObjMap"
`;

exports[`gRPC Reflection Example should get greeting correctly: greeting-result 1`] = `
{
  "data": {
    "helloworld_GreetingService_GetGreeting": {
      "greeting": "Hello World",
    },
  },
}
`;
